Initialize a `Grid` object with a given `Fuse` limit

tiyash-basu-frequenz opened the pull request:
Calling `microgrid.initialize()` now also initializes the microgrid's grid connection point as a singleton object of a newly added type `Grid`. This object can be obtained by calling `microgrid.grid.get()`. This object exposes the max current that can course through the grid connection point, which is useful for the power distribution algorithm. The max current is provided by the Microgrid API, and can be obtained by calling `microgrid.grid.get().fuse.max_current`.

Note that a microgrid is allowed to have zero or one grid connection point. Microgrids configured as islands will have zero grid connection points, and microgrids configured as grid-connected will have one grid connection point.

A new class `Fuse` has been added to represent fuses. This class has a member variable `max_current` which represents the maximum current that can course through the fuse. If the current flowing through a fuse is greater than this limit, then the fuse will break the circuit.

tiyash-basu-frequenz committed:
Add a module for grid connection points

Grid connection points are nodes in the component graph that abstract
grids in a microgrid. Such nodes have their own metadata, which include
their rated fuse limits. Microgrids with zero or one grid connection points
are accepted. Note that a microgrid may have zero grid connection points,
if it is configured as an island.

This commit adds a module exposing a representation of such grid connection
points. The module is named `frequenz.sdk.microgrid.grid`.

The `grid` module exposes a singleton object that represents the grid
connection point. It gets initialized when `microgrid.initialize()` is
called. It can be obtained by calling `microgrid.grid_connection.get()`.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Add metadata to Components

This commit introduces metadata to Components by incorporating
GridMetadata, which now includes the maximum grid current,
specified in Amperes.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Allow setting grid metadata in mocks

This commit allows tests to specify the `rated_fuse_current` member in
the metadata objects of grid connection points. The test data supplied
using this mechanism can be used in downstream validations.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Add a Fuse class and use it in Grid

This commit introduces a class to represent fuses. Fuses have just one
property representing their rated max current. If the current flowing
through a fuse is greater than this limit, then the fuse will break the
circuit.

This commit also introduces using a Fuse object to represent the allowed
max current at a grid connection point.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Use fuse in GridMetadata

This commit uses a Fuse object to represent the max current limit in
component metadata in the internal layer that parses component objects
obtained directly from the Microgrid API. This makes the abstraction of
fuses one level deeper, increasing the chances of the `grid_connection.py`
file being more stable in the future if/when further changes are
introduced in the Microgrid API.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Implement a hash method for `Component`

Objects of the `Component` type are used in `ComponentGraph` in sets. The
implicit hash method works there for now, but adding new members to the
`Component` class may make it unhashable in the future, if the added
members themselves are unhashable. Implementing an explicit hash method
here removes this burden from from future member types.

Component IDs are supposed to be unique in a microgrid, therefore hashing a
`Component` object just based on its component ID is enough.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz commented on src/frequenz/sdk/microgrid/component/_component.py:
```diff
@@ -106,13 +108,64 @@ def _component_category_from_protobuf(
+class GridMetadata(ComponentMetadata):
+    """Metadata for a grid connection point."""
+
+    def __init__(self, max_current: float) -> None:
```
It'd be nice if this were a dataclass. But changing it to a dataclass turns `Component.metadata` into a dict, causing a few comparison tests to fail. Has anyone here encountered such an issue before?

daniel-zullo-frequenz replied:
Have you tried with `@dataclass(frozen=True)`?

matthias-wende-frequenz replied:
Is there a reason why `max_current` is not of type `Current`?

tiyash-basu-frequenz replied:
>Have you tried with @dataclass(frozen=True)?

yes, but it did not work.

tiyash-basu-frequenz replied:
>Is there a reason why max_current is not of type Current?

Because that's how I started, and it did not make sense to change it to `Current` because this gets changed to `Fuse` anyway.

tiyash-basu-frequenz replied:
If there are no further suggestions, then I'll mark this resolved.

llucax replied:
> yes, but it did not work.

How it didn't work? Was there some error? It would be nice to keep the same style as other classes and besides immutability, dataclass will also take care of implementing `__eq__` for example.

tiyash-basu-frequenz replied:
Here is an example with `fuse`:
```
tests/microgrid/test_graph.py::Test_MicrogridComponentGraph::test_refresh_from_api <actor_restart_limit> Changing the restart limit from None to 0
FAILED<actor_restart_limit> Resetting restart limit to None
...
E       AssertionError: assert
{Component(component_id=131, category=<ComponentCategory.EV_CHARGER: 6>, type=None, metadata=None), Component(component_id=101, category=<ComponentCategory.GRID: 1>, type=None, metadata={'fuse': {'max_current': Current(value=0, exponent=0)}}), Component(component_id=111, category=<ComponentCategory.METER: 2>, type=None, metadata=None)}
== {Component(component_id=131, category=<ComponentCategory.EV_CHARGER: 6>, type=None, metadata=None), Component(component_id=101, category=<ComponentCategory.GRID: 1>, type=None, metadata=GridMetadata(fuse=Fuse(max_current=Current(value=0, exponent=0)))), Component(component_id=111, category=<ComponentCategory.METER: 2>, type=None, metadata=None)}
```

The LHS has a dict, but the RHS has an object. The method of constructing these two look the same to me, but I am definitely missing something.

tiyash-basu-frequenz replied:
Found the culprit: https://github.com/frequenz-floss/frequenz-sdk-python/blob/1fe5cb8c0103f37580988fa03da3a345a3079290/src/frequenz/sdk/microgrid/_graph.py#L476

The graph module is converting components to dicts.

Does anyone know why this conversion to a dict is required?
Edit: because the `add_node` method takes only one positional argument, which is the component ID. Everything after it therefore has to be expanded.

tiyash-basu-frequenz replied:
Will push changes today to get this as a dataclass.

tiyash-basu-frequenz replied:
Done. Ready for a look.

daniel-zullo-frequenz commented on src/frequenz/sdk/microgrid/fuse.py:
```diff
@@ -0,0 +1,19 @@
+    """Fuse data class.
+
+    Attributes:
+        max_current: Rated current of the fuse, in Amperes.
```
As `max_current` is `Current` type we don't need to explicitly document the unit.
And we usually document attributes below the attribute definition as follows:

```python
@dataclass(frozen=True)
class Fuse:
    """Fuse data class."""

    max_current: Current
    """Rated current of the fuse."""

tiyash-basu-frequenz replied:
Done! Thanks for the suggestion. Do we want to enforce this style in doc tests?

daniel-zullo-frequenz replied:
No, not really. Test might not need documentation except some special cases, it is up to the developer

daniel-zullo-frequenz commented on src/frequenz/sdk/microgrid/grid.py:
```diff
@@ -0,0 +1,82 @@
+        fuse: The fuse protecting the grid connection point.
+    """
+
+    fuse: Fuse
```
```suggestion
    """A grid connection point."""

    fuse: Fuse
    """The fuse protecting the grid connection point."""
```

tiyash-basu-frequenz replied:
Done

daniel-zullo-frequenz commented on src/frequenz/sdk/microgrid/grid.py:
```diff
@@ -0,0 +1,82 @@
+        )
+    elif grid_connections_count > 1:
+        raise RuntimeError(
+            f"Expected at most one grid connection, got {len(grid_connections)}"
```
Alternatively you can use the dedicated variable `grid_connections_count`

tiyash-basu-frequenz replied:
Done

daniel-zullo-frequenz commented on src/frequenz/sdk/microgrid/grid.py:
```diff
@@ -0,0 +1,82 @@
+
+    Raises:
+        RuntimeError: If there is more than 1 grid connection point in the
+            microgrid.
```
Actually there are 3 cases where `RuntimeError` is raised that you might want briefly to mention them

tiyash-basu-frequenz replied:
Done

daniel-zullo-frequenz commented on src/frequenz/sdk/microgrid/grid.py:
```diff
@@ -0,0 +1,82 @@
+        )
+    else:
+        if grid_connections[0].metadata is None:
+            raise RuntimeError("Grid metadata is None")
```
Optionally a more "pythonic" way to do it is `if not grid_connections[0].metadata:`

tiyash-basu-frequenz replied:
Same: https://github.com/frequenz-floss/frequenz-sdk-python/pull/610#discussion_r1307389436

> Oof that's a weird one. It feels like casting fuse to a bool type.

daniel-zullo-frequenz commented on src/frequenz/sdk/microgrid/grid.py:
```diff
@@ -0,0 +1,82 @@
+
+        fuse = grid_connections[0].metadata.fuse
+
+        if fuse is None:
```
```suggestion
        if not fuse:
```

tiyash-basu-frequenz replied:
Oof that's a weird one. It feels like casting fuse to a `bool` type.

daniel-zullo-frequenz replied:
yeah feel free to keep it as it currently is

daniel-zullo-frequenz commented on tests/microgrid/mock_api.py:
```diff
@@ -100,6 +102,17 @@ def add_component(
             )
+        elif (
+            component_category == ComponentCategory.COMPONENT_CATEGORY_GRID
+            and max_current is not None
```
if `max_current` should be always different than zero then I'd suggest to use `and max_current` instead as this covers the check for not `None` and not `0`

tiyash-basu-frequenz replied:
> if `max_current` should be always different than zero

I did not make that assumption here. I think a mock should not have conditions on data values. The functional implementation should have it instead.

But it raises a good point. In this case, would it make sense to raise an error in the `Grid` module if `fuse.max_current == 0`, given the low-level layer it exists in? Or should it be handled by the power distributor?

daniel-zullo-frequenz replied:
> I did not make that assumption here. I think a mock should not have conditions on data values. The functional implementation should have it instead.

You're right, agreed.

> would it make sense to raise an error in the Grid module if fuse.max_current == 0, given the low-level layer it exists in? Or should it be handled by the power distributor?

I think It does make sense to do it in the Grid module to prevent programming or setups errors. And additionally can be done in the power distributor. Maybe @matthias-wende-frequenz has another opinion on this matter?

matthias-wende-frequenz replied:
> has another opinion on this matter?

No other opinion.

tiyash-basu-frequenz replied:
Okay, then I think we can resolve this discussion here, and pass this on to the power distributor.

daniel-zullo-frequenz reviewed (commented):
I have a few minor suggestions to check for and LGTM otherwise

daniel-zullo-frequenz commented on tests/microgrid/mock_api.py:
```diff
@@ -89,6 +90,7 @@ def add_component(
         self,
         component_id: int,
         component_category: ComponentCategory.V,
+        max_current: Optional[int] = None,
```
Also I wonder if `max_current` should be `Optional[Current]` type instead

tiyash-basu-frequenz replied:
This gets changed to `Fuse` a few commits later, so this change is probably not worth the effort.

daniel-zullo-frequenz replied:
Are you sure? I'm checking all files changes (not per commit) and this is still there :/ Maybe it's an issue in my browser :thinking:

tiyash-basu-frequenz replied:
Oh this is the `mock_api` class. Sorry, you are right, I misread the filename. Will change it to `Current`.

tiyash-basu-frequenz replied:
Done

daniel-zullo-frequenz commented on tests/microgrid/test_graph.py:
```diff
@@ -816,8 +823,11 @@ async def test_refresh_from_api(self) -> None:
         servicer.set_connections([(101, 111), (111, 131)])
         await graph.refresh_from_api(client)
 
+        grid_max_current = Current.from_amperes(0.0)
```
We can use the quantity zero constructor instead as follows:
```suggestion
        grid_max_current = Current.zero()
```

tiyash-basu-frequenz replied:
Nice! Done

matthias-wende-frequenz commented on src/frequenz/sdk/microgrid/grid.py:
```diff
@@ -0,0 +1,75 @@
+            connection point, in Amperes.
+    """
+
+    max_current: float
```
This should be of type `Current`.

llucax commented on RELEASE_NOTES.md:
```diff
@@ -92,6 +92,12 @@ This release replaces the `@actor` decorator with a new `Actor` class.
 - `Actor`: This new class inherits from `BackgroundService` and it replaces the `@actor` decorator.
 
+- Calling `microgrid.initialize()` now also initializes the microgrid's grid connection point as a singleton object of a newly added type `Grid`. This object can be obtained by calling `microgrid.grid.get()`. This object exposes the max current that can course through the grid connection point, which is useful for the power distribution algorithm. The max current is provided by the Microgrid API, ...
+  Note that a microgrid is allowed to have zero or one grid connection point. Microgrids configured as islands will have zero grid connection points, and microgrids configured as grid-connected will have one grid connection point.
```
For proper formatting, an empty line should be left if you want both paragraphs to be really rendered as paragraphs.

```suggestion

  Note that a microgrid is allowed to have zero or one grid connection point. Microgrids configured as islands will have zero grid connection points, and microgrids configured as grid-connected will have one grid connection point.
```

tiyash-basu-frequenz replied:
Done

llucax commented on src/frequenz/sdk/microgrid/component/_component.py:
```diff
@@ -106,13 +112,78 @@ def _component_category_from_protobuf(
+
+        Returns:
+            The fuse associated with this component.
+        """
```
Any reason not to provide a default implementation returning `None` here if having a `fuse` is optional? (so `fuse` won't be an abstract method). Even more, it could be a dataclass with just an optional property fuse:

```py
@dataclass(froze=True)
class ComponentMetadata(ABC):
    """Base class for component metadata classes."""

    fuse: Fuse | None = None
        """The fuse associated with this component."""
```

Anyone can override it if they want.

tiyash-basu-frequenz replied:
Added a default that returns `None`

llucax commented on src/frequenz/sdk/microgrid/component/_component.py:
```diff
@@ -106,13 +112,78 @@ def _component_category_from_protobuf(
+        self._fuse = fuse
+
+    @property
+    def fuse(self) -> Optional[Fuse]:
```
Nitpick: *modern* Python uses `Fuse | None` (but we are trying to modernize the code base, so it would be nice if new code uses the new style, although if the file in general is using the old style it is OK to leave it and make the modernization of the whole file in another commit/PR)

```suggestion
    def fuse(self) -> Fuse | None:
```

tiyash-basu-frequenz replied:
Amazing! I inserted a commit to modernize the declarations, and then changed my commits to use this style.

tiyash-basu-frequenz committed:
Refactor optional type declarations in _component.py

This commit refactors the declarations of types in format `Optional[X]` to
`X | None`.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz commented:
Rebased on latest commit in v0.x.x.

llucax commented on src/frequenz/sdk/microgrid/grid.py:
```diff
@@ -0,0 +1,80 @@
+    """The fuse protecting the grid connection point."""
+
+
+_GRID: Optional[Grid] = None
```
```suggestion
_GRID: Grid | None = None
```

tiyash-basu-frequenz replied:
Done

llucax commented on src/frequenz/sdk/microgrid/grid.py:
```diff
@@ -0,0 +1,80 @@
+
+import logging
+from dataclasses import dataclass
+from typing import Iterable, Optional
```
New import location for `Iterable` (and `Optional` is not necessary)

```suggestion
from collections.abc import Iterable
```

tiyash-basu-frequenz replied:
Done

matthias-wende-frequenz replied:
~isn't typing the new location?~ Forget it. Luca was right and chatgpt was wrong :D

tiyash-basu-frequenz replied:
Lemme see what pylance suggests.

tiyash-basu-frequenz replied:
pylance suggests `typing`. I changed it to `typing`.

llucax replied:
Sorry, but `pylance` is wrong :)

Python provided the stuff in `collections.abc` but at first it could just use regular types as type hints, so they needed to add special types to the `typing` module (there was even a `typing.Int`), but then it got the ability to use regular types, the use of types in `typing` become deprecated.

tiyash-basu-frequenz replied:
reverted

llucax replied:
Thanks! Just for more context:

![image](https://github.com/frequenz-floss/frequenz-sdk-python/assets/1031485/b9f99c93-0f37-4ded-bec7-d62be0be2389)

llucax commented on src/frequenz/sdk/microgrid/grid.py:
```diff
@@ -0,0 +1,80 @@
+        _GRID = Grid(fuse)
+
+
+def get() -> Optional[Grid]:
```
```suggestion
def get() -> Grid | None:
```

tiyash-basu-frequenz replied:
~I think it's better if we stay consistent in this file and use the `Optional[X]` format here for now. We should, however, update occurrences of this format everywhere in the SDK. Although, that sounds out-of-scope for this PR to me.~

Sorry, please ignore the previous text. Done

llucax commented on tests/microgrid/mock_api.py:
```diff
@@ -89,6 +92,7 @@ def add_component(
         self,
         component_id: int,
         component_category: ComponentCategory.V,
+        max_current: Optional[Current] = None,
```
```suggestion
        max_current: Current | None = None,
```

tiyash-basu-frequenz replied:
I think it's better if we stay consistent in this file and use the `Optional[X]` format here for now. We should, however, update occurrences of this format everywhere in the SDK. Although, that sounds out-of-scope for this PR to me.

llucax reviewed (dismissed):
A few more optional (no pun intended) comments. Maybe @shsms should have a quick look  too as he is probably the most familiar with the microgrid API code.
